Regression: `lto = true` in build.rs causes `cargo test --release --doc` to fail to link

This regression occurs in the beta:

**Problem**

`cargo test --doc --release` fails with linker error "fatal error LNK1107: invalid or corrupt file: cannot read at 0xD2A" on Windows using MSVC 2019 when `lto = true` is in the release build profile, in the beta and nightly channels.  `cargo test --doc` (non-release builds) succeeds. `cargo test --doc --release` succeeds in the current stable release, `cargo 1.45.1 (f242df6ed 2020-07-22)`.

Cargo.toml:

```
[package]
name = "lto-rustdoc"
version = "0.1.0"
edition = "2018"

[profile.release]
opt-level = 3
lto = true
```

src/lib.rs:

```
/// ```
/// lto_rustdoc::foo();
/// ```
pub fn foo() {}
```

The failures do not occur when the current stable cargo is used.

**Possible Solution(s)**

I suspect that PR #8192 or one of the follow-ups such as PR #8349 caused this regression.

**Notes**

```
$ cargo +beta version
cargo 1.47.0-beta (51b66125b 2020-08-19)
```

This also occurs in Nightly. It has been occurring in nightly for several weeks, and in the last few weeks it has occurred in Beta. Thus, the regression most likely happened before the last promotion of Nightly to Beta.

I used Process Monitor to capture the command lines for Cargo 1.45.1:
```
"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.27.29110\bin\HostX64\x64\link.exe"
/NOLOGO
/NXCOMPAT
/LIBPATH:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib
C:\Users\b\AppData\Local\Temp\rustdoctestiHYAmZ\rust_out.rust_out.7rcbfp3g-cgu.0.rcgu.o
/OUT:C:\Users\b\AppData\Local\Temp\rustdoctestiHYAmZ\rust_out
C:\Users\b\AppData\Local\Temp\rustdoctestiHYAmZ\rust_out.33dyzt1ekirinwy8.rcgu.o
/OPT:REF,NOICF
/DEBUG
/NATVIS:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\intrinsic.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\liballoc.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\libcore.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\etc\libstd.natvis
/LIBPATH:P:\testcases\lto-rustdoc\target\release\deps
/LIBPATH:P:\testcases\lto-rustdoc\target\release\deps
/LIBPATH:C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib
P:\testcases\lto-rustdoc\target\release\deps\liblto_rustdoc-5744db2bc717fc58.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libstd-098718ea823008f4.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libpanic_unwind-6a7fb52d1138932f.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libhashbrown-6a663e98da9bc9d7.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_std_workspace_alloc-35acfeb1bd722cd7.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libbacktrace-5789469546c93c74.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_demangle-6b40c36b7a5ce598.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libunwind-ce5287493729af9a.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcfg_if-3d66b8897882cc1e.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\liblibc-2f1bb63f49853f4b.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\liballoc-e769bcc0ed78b3c4.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_std_workspace_core-2182f71ae05a8995.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcore-77514af2604d16ea.rlib
C:\Users\b\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcompiler_builtins-ebafcb87dc3d8679.rlib
advapi32.lib
ws2_32.lib
userenv.lib
msvcrt.lib
```

and the current beta:

```
"C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.27.29110\bin\HostX64\x64\link.exe"
/NOLOGO
/NXCOMPAT
/LIBPATH:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib
C:\Users\b\AppData\Local\Temp\rustdoctestNeUiEv\rust_out.rust_out.7rcbfp3g-cgu.0.rcgu.o
/OUT:C:\Users\b\AppData\Local\Temp\rustdoctestNeUiEv\rust_out
C:\Users\b\AppData\Local\Temp\rustdoctestNeUiEv\rust_out.33dyzt1ekirinwy8.rcgu.o
/OPT:REF,NOICF
/DEBUG
/NATVIS:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\etc\intrinsic.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\etc\liballoc.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\etc\libcore.natvis
/NATVIS:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\etc\libstd.natvis
/LIBPATH:P:\testcases\lto-rustdoc\target\release\deps
/LIBPATH:P:\testcases\lto-rustdoc\target\release\deps
/LIBPATH:C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib
P:\testcases\lto-rustdoc\target\release\deps\liblto_rustdoc-c5f7efb3b772d87e.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libstd-f5b401742159082b.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libpanic_unwind-e179a42e3654f4e1.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_demangle-6606b5facb12279a.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libhashbrown-1a0e7b07a7e9bb53.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_std_workspace_alloc-9aeb694399775ad0.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libunwind-4fc2d77659a37b5c.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcfg_if-8e0f5ece44f1eaea.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\liblibc-fad582f67483dcdf.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\liballoc-0f8a186f598b0388.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\librustc_std_workspace_core-ec163dda744bc8b6.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcore-6d2170587abafa35.rlib
C:\Users\b\.rustup\toolchains\beta-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib\libcompiler_builtins-e07801ab81cac207.rlib
advapi32.lib
ws2_32.lib
userenv.lib
msvcrt.lib
```

The command lines seem to be the same.

The same problem occurs on Linux too:

```
 note: /home/travis/build/briansmith/ring/target/x86_64-unknown-linux-gnu/release/deps/libring-a6aeb6a391324468.rlib: error adding symbols: File format not recognized
          collect2: error: ld returned 1 exit status
```

--- a/src/cargo/core/compiler/context/mod.rs
+++ b/src/cargo/core/compiler/context/mod.rs
@@ -210,7 +210,8 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
             // Collect information for `rustdoc --test`.
             if unit.mode.is_doc_test() {
                 let mut unstable_opts = false;
-                let args = compiler::extern_args(&self, unit, &mut unstable_opts)?;
+                let mut args = compiler::extern_args(&self, unit, &mut unstable_opts)?;
+                args.extend(compiler::lto_args(&self, unit));
                 self.compilation.to_doc_test.push(compilation::Doctest {
                     unit: unit.clone(),
                     args,

--- a/src/cargo/core/compiler/mod.rs
+++ b/src/cargo/core/compiler/mod.rs
@@ -797,28 +797,9 @@ fn build_base_args(
         cmd.arg("-C").arg(format!("panic={}", panic));
     }
 
-    match cx.lto[unit] {
-        lto::Lto::Run(None) => {
-            cmd.arg("-C").arg("lto");
-        }
-        lto::Lto::Run(Some(s)) => {
-            cmd.arg("-C").arg(format!("lto={}", s));
-        }
-        lto::Lto::Off => {
-            cmd.arg("-C").arg("lto=off");
-        }
-        lto::Lto::ObjectAndBitcode => {} // this is rustc's default
-        lto::Lto::OnlyBitcode => {
-            cmd.arg("-C").arg("linker-plugin-lto");
-        }
-        lto::Lto::OnlyObject => {
-            cmd.arg("-C").arg("embed-bitcode=no");
-        }
-    }
+    cmd.args(&lto_args(cx, unit));
 
     if let Some(n) = codegen_units {
-        // There are some restrictions with LTO and codegen-units, so we
-        // only add codegen units when LTO is not used.
         cmd.arg("-C").arg(&format!("codegen-units={}", n));
     }
 
@@ -944,6 +925,23 @@ fn build_base_args(
         }
     }
     Ok(())
+}
+
+fn lto_args(cx: &Context<'_, '_>, unit: &Unit) -> Vec<OsString> {
+    let mut result = Vec::new();
+    let mut push = |arg: &str| {
+        result.push(OsString::from("-C"));
+        result.push(OsString::from(arg));
+    };
+    match cx.lto[unit] {
+        lto::Lto::Run(None) => push("lto"),
+        lto::Lto::Run(Some(s)) => push(&format!("lto={}", s)),
+        lto::Lto::Off => push("lto=off"),
+        lto::Lto::ObjectAndBitcode => {} // this is rustc's default
+        lto::Lto::OnlyBitcode => push("linker-plugin-lto"),
+        lto::Lto::OnlyObject => push("embed-bitcode=no"),
+    }
+    result
 }
 
 fn build_deps_args(

--- a/src/cargo/core/profiles.rs
+++ b/src/cargo/core/profiles.rs
@@ -299,7 +299,7 @@ impl Profiles {
             let release = matches!(self.requested_profile.as_str(), "release" | "bench");
 
             match mode {
-                CompileMode::Test | CompileMode::Bench => {
+                CompileMode::Test | CompileMode::Bench | CompileMode::Doctest => {
                     if release {
                         (
                             InternedString::new("bench"),
@@ -312,10 +312,7 @@ impl Profiles {
                         )
                     }
                 }
-                CompileMode::Build
-                | CompileMode::Check { .. }
-                | CompileMode::Doctest
-                | CompileMode::RunCustomBuild => {
+                CompileMode::Build | CompileMode::Check { .. } | CompileMode::RunCustomBuild => {
                     // Note: `RunCustomBuild` doesn't normally use this code path.
                     // `build_unit_profiles` normally ensures that it selects the
                     // ancestor's profile. However, `cargo clean -p` can hit this

--- a/tests/testsuite/lto.rs
+++ b/tests/testsuite/lto.rs
@@ -1,6 +1,6 @@
 use cargo::core::compiler::Lto;
 use cargo_test_support::registry::Package;
-use cargo_test_support::{project, Project};
+use cargo_test_support::{basic_manifest, project, Project};
 use std::process::Output;
 
 #[cargo_test]
@@ -514,16 +514,19 @@ fn cdylib_and_rlib() {
     p.cargo("test --release -v --manifest-path bar/Cargo.toml")
         .with_stderr_unordered(
             "\
-[FRESH] registry v0.0.1
-[FRESH] registry-shared v0.0.1
+[COMPILING] registry v0.0.1
+[COMPILING] registry-shared v0.0.1
+[RUNNING] `rustc --crate-name registry [..]-C embed-bitcode=no[..]
+[RUNNING] `rustc --crate-name registry_shared [..]-C embed-bitcode=no[..]
 [COMPILING] bar [..]
+[RUNNING] `rustc --crate-name bar [..]--crate-type cdylib --crate-type rlib [..]-C embed-bitcode=no[..]
 [RUNNING] `rustc --crate-name bar [..]-C embed-bitcode=no --test[..]
 [RUNNING] `rustc --crate-name b [..]-C embed-bitcode=no --test[..]
 [FINISHED] [..]
-[RUNNING] [..]
-[RUNNING] [..]
+[RUNNING] [..]target/release/deps/bar-[..]
+[RUNNING] [..]target/release/deps/b-[..]
 [DOCTEST] bar
-[RUNNING] `rustdoc --crate-type cdylib --crate-type rlib --test [..]
+[RUNNING] `rustdoc --crate-type cdylib --crate-type rlib --test [..]-C embed-bitcode=no[..]
 ",
         )
         .run();
@@ -627,7 +630,7 @@ fn test_profile() {
 [COMPILING] bar v0.0.1
 [RUNNING] `rustc --crate-name bar [..]crate-type lib[..]
 [COMPILING] foo [..]
-[RUNNING] `rustc --crate-name foo [..]--crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no[..]
+[RUNNING] `rustc --crate-name foo [..]--crate-type lib --emit=dep-info,metadata,link -C linker-plugin-lto[..]
 [RUNNING] `rustc --crate-name foo [..]--emit=dep-info,link -C lto=thin [..]--test[..]
 [FINISHED] [..]
 [RUNNING] [..]
@@ -680,12 +683,66 @@ fn dev_profile() {
 [COMPILING] bar v0.0.1
 [RUNNING] `rustc --crate-name bar [..]crate-type lib[..]
 [COMPILING] foo [..]
-[RUNNING] `rustc --crate-name foo [..]--crate-type lib --emit=dep-info,metadata,link -C linker-plugin-lto [..]
+[RUNNING] `rustc --crate-name foo [..]--crate-type lib --emit=dep-info,metadata,link -C embed-bitcode=no [..]
 [RUNNING] `rustc --crate-name foo [..]--emit=dep-info,link -C embed-bitcode=no [..]--test[..]
 [FINISHED] [..]
 [RUNNING] [..]
 [DOCTEST] foo
 [RUNNING] `rustdoc [..]
 ")
+        .run();
+}
+
+#[cargo_test]
+fn doctest() {
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.1.0"
+                edition = "2018"
+
+                [profile.release]
+                lto = true
+
+                [dependencies]
+                bar = { path = "bar" }
+            "#,
+        )
+        .file(
+            "src/lib.rs",
+            r#"
+                /// Foo!
+                ///
+                /// ```
+                /// foo::foo();
+                /// ```
+                pub fn foo() { bar::bar(); }
+            "#,
+        )
+        .file("bar/Cargo.toml", &basic_manifest("bar", "0.1.0"))
+        .file(
+            "bar/src/lib.rs",
+            r#"
+                pub fn bar() { println!("hi!"); }
+            "#,
+        )
+        .build();
+
+    p.cargo("test --doc --release -v")
+        .with_stderr_contains("[..]`rustc --crate-name bar[..]-C embed-bitcode=no[..]")
+        .with_stderr_contains("[..]`rustc --crate-name foo[..]-C embed-bitcode=no[..]")
+        // embed-bitcode should be harmless here
+        .with_stderr_contains("[..]`rustdoc [..]-C embed-bitcode=no[..]")
+        .run();
+
+    // Try with bench profile.
+    p.cargo("test --doc --release -v")
+        .env("CARGO_PROFILE_BENCH_LTO", "true")
+        .with_stderr_contains("[..]`rustc --crate-name bar[..]-C linker-plugin-lto[..]")
+        .with_stderr_contains("[..]`rustc --crate-name foo[..]-C linker-plugin-lto[..]")
+        .with_stderr_contains("[..]`rustdoc [..]-C lto[..]")
         .run();
 }

